library/std/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,53 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
```

Comment:
@BatmanAoD at least I hope this is true for all our `*-unwind` ABIs...

Reply:
I think it would always be UB if, with `panic=unwind`, a non-unwinding function pointer were used to invoke an `-unwind` function and the callee actually panicked (without catching it). But otherwise yes, the calling convention and other ABI characteristics should be identical.

Reply:
> I think it would always be UB if, with panic=unwind, a non-unwinding function pointer were used to invoke an -unwind function and the callee actually panicked (without catching it). 

Right, here we're actually saying that's UB no matter what.

The more critical point is that the docs say that it is *allowed* to call a `"C"` function with a `"C-unwind"` function pointer.

Reply:
That should definitely be allowed, yes. 

I believe it would only be a problem for implementation if LLVM IR (or IR for some other hypothetical backend) declared `noexcept` to be part of the calling convention, which I don't think it ever would (it certainly doesn't today).